### ISSUE TITLE
introduce java.io.Serializable to all Ceylon classes

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-compiler#2329]
Since Java EE infrastructure, and other frameworks, often depend upon a class being serializable, we should make every generated class that directly extends `Basic` or `Object` implement `java.io.Serializable`.

This doesn't apply to `object`s, which wouldn't be serializable. Note that this works in concert with #2324, but is not dependent on #2324.

See sgalles/ceylon-dddsample#1.

**Comment by quintesse** [comment]
But can we actually do that? I mean I guess there must be reasons Java classes don't all have this interface by default, right?

**Comment by gavinking** [comment]
No particularly good reasons, no.

**Comment by tombentley** [comment]
For the serialization to actually work in all but the most trivial of circumstances we will need:

* To make some of the helper classes in the runtime Serializable too (`TypeDescriptor`, the variable boxes, AbstractIterable, AbstractIterator, AbstractCallable) 
* To make companion classes Serializable (or make those members `transient`, and lazily initialized)
* A lot of testing

That's quite a lot of work, actually.

**Comment by gavinking** [comment]
1. Right now I don't care if function refs, anonymous functions, nested classes, streams, or comprehensions are serializable. That's something we could come back to later if necessary.
2. Given that, I don't think it's actually necessary for boxes and iterable superclasses to be serializable&mdash;however, I'm marking the `Serializable` right now, just in case.
3. I agree that `TypeDescriptor` must be serializable. I will take a look to see if that's a problem. If not, I will make it so.
4. We indeed need to do something about companion classes, which would require what looks like a very modest change to the compiler. The trivial fix is to mark them `Serializable`. Alternative fixes include lazily initializing them or initializing them in a `readObject()` method.

This does _not_ call for a lot of testing. If we can serialize a trivial class with an `ArrayList` and a `HashMap`, then we're already way ahead of where we are today.

**Comment by gavinking** [comment]
Boxes and `TypeDescriptor`s are now serializable.

**Comment by tombentley** [comment]
Thanks

**Comment by tombentley** [comment]
Done. I added a `readResolve()` for toplevel `object` declarations, but toplevel classes with all value constructors (no callable constructors) are not serializable, since that would suddenly allow non-identical duplicates. We'd need a `writeReplace()` to fix that. 

I'm sure there are other corner cases too, but it should be good enough for now.

**Comment by gavinking** [comment]
Awesome!

**Comment by FroMage** [comment]
Why don't we only do this for ceylon serializable (annotated as such) classes? Or don't we require an annotation, I haven't checked.

**Comment by gavinking** [comment]
Well the question is whether the `serializable` annotation actually has any intersection in terms of semantics with java serializability, which is actually quite different.

**Comment by sgalles** [comment]
@tombentley FYI I noticed that if a Ceylon class is still explicitly marked with `java.io.Serializable` in the code, it causes an error that somewhat fuzzy : 
`repeated interface` without line number, and without the name of the interface.

**Comment by tombentley** [comment]
Ah, yes thanks @sgalles that was something I meant to test and then forgot about, let me sort that out now...

**Comment by tombentley** [comment]
Now fixed.

**Comment by FroMage** [comment]
> Well the question is whether the serializable annotation actually has any intersection in terms of semantics with java serializability, which is actually quite different.

Well, intuitively I'd say it should, but I see that `Integer` is not `serializable`, so apparently our annotation means something else?

**Comment by tombentley** [comment]
Well our `serializable` is about having (being able to obtain) a recipe to reconstruct the instance. If there was a meaningful way to reconstruct an `Integer` then it would be serializable, but there isn't because it's so primitive. We basically assume that the serialization format comes with the ability to encode an Integer some how, and that it can therefore supply the value of a particular Integer at deserialization time.

Java doesn't have this issue because `long`, `double` etc, being primitive, can't satisfy Serializable, but their serialization format obviously knows how to serialize them anyway.

**Comment by tombentley** [comment]
TBH, I'm a little dubious about this whole implicit `Serialiable` thing too. Because it's implicit we're effectively saying the compiler can do all the hard work, and the user doesn't have any knobs to twiddle to actually make it work when it turns out the compiler can't always do a good enough job. Obviously we plan to have `transient` and some kind of `serialVersionUid`, but I can see a long list of issues where we get it wrong, or suboptimal and have to fix it, or add knobs for the user to get what they want.

**Comment by FroMage** [comment]
OK so you're saying our `Integer` is _implicitly_ `serializable`, right? Why not make it explicit then? Any good reason?

What you're saying seems to also indicate that only making things annotated with `serializable` be `java.io.Serializable` would work, right? Except perhaps objects which can't be annotated?

**Comment by tombentley** [comment]
> OK so you're saying our `Integer` is *implicitly* `serializable`, right? Why not make it explicit then? Any good reason?

Well until now `serializable` has been used as an instruction to the backend to generate the necessary code to be able to obtain the serializable state of an object and to be able to create and then initialize an instance with that state. Very clear semantics. Now you want to add that annotation to some classes, whose state cannot be expressed in terms of their fields (because they have no fields!), and the compiler somehow needs to know not to add those extra methods. (I know that these things are native, so the compiler never actually sees them, but you get the point). Suddenly I don't know what `serializable` is supposed to mean, and whose responsibility it is to ensure that the class can be serialized. 

Maybe a better word than `serializable` would be `decomposable`, because that's what it means right now.

> What you're saying seems to also indicate that only making things annotated with serializable be java.io.Serializable would work, right? Except perhaps objects which can't be annotated?

Well consider top level `object` declarations. They can all be serialized (in the Java sense) because they shouldn't actually be instantiated in the deserializing VM, but instead the value of the global within that VM should be substituted in the object graph. We do that with a `readResolve()` and it works. What you're advocating is that they're only `Serializble` if people explicitly annotate them `serializable`, which they're just not going to do all the time. So the whole thing because less useful.

Now in our serialization API `object`s do not need to be `serializable`, but the serialization library can support object graphs with `object`s by serializing a "reference" and resolving that at deserialization time (basically saying "the thing with this ID is this instance which I'm giving you", instead of "the thing with this ID is an instance of class C which I expect you to instantiate for me").

Basically the semantics are different.

**Comment by gavinking** [comment]
> the user doesn't have any knobs to twiddle to actually make it work when it turns out the compiler can't always do a good enough job

Well the user can always implement `readObject()`/`writeObject()`.

**Comment by gavinking** [comment]
@tombentley what you just wrote lines up pretty well with my conception of what `serializable` means to us.

**Comment by FroMage** [comment]
> Suddenly I don't know what serializable is supposed to mean, and whose responsibility it is to ensure that the class can be serialized.

Well, we already have special cases for `Integer` and friends, right? So that's just keeping them special.

But OK, what you're saying is that we can even serialise things not annotated `serializable`. But TBH I don't understand then why anyone would use that annotation?

**Comment by gavinking** [comment]
@FroMage the `serializable` annotation is an instruction to the compiler to generate a certain kind of code that lets an object be used with a certain API. `Integer` and `Float` can't be passed to that API.

In fact, that's an argument that `Serializable` should in fact be an interface you implement.

**Comment by FroMage** [comment]
Guys, you're going to have to start explaining better what things like "to generate a certain kind of code that lets an object be used with a certain API."  mean ;)

Hell, I'd like to read about our serialisation, I realise I've no clue what it does.

**Comment by quintesse** [comment]
Well the interface would be a good idea anyway I guess because I'm assuming that at some point we'd want "serialization-by-proxy" where you could register an object that can serialize objects that don't know how the serialize themselves or that are too costly/difficult to serialize directly (eg. most collection classes).

**Comment by gavinking** [comment]
OK, so, `serializable` lines up very closely to the notion of `@Entity` in something like JPA, or similar annotations you find in XML/json serialization frameworks. `Integer`s and `String`s aren't "entities", and they don't have an identity separate from the object to which they belong. `serializable` objects do indeed have such an identity.

**Comment by gavinking** [comment]
> Well the interface would be a good idea anyway

Actually it wouldn't really be that good of an idea, since, like in Java, it would guarantee almost nothing. A subclass of a serializable class can still be unserializable, and it's almost impossible to force it to be serializable.

`java.io.Serializable` is a totally useless contract that promises essentially nothing, and defines no semantics that can be enforced at compile time.

**Comment by gavinking** [comment]
>  I'm assuming that at some point we'd want "serialization-by-proxy" where you could register an object that can serialize objects that don't know how the serialize themselves or that are too costly/difficult to serialize directly (eg. most collection classes).

"at some point"? WTF, we already have this. That's the whole point of this.

**Comment by quintesse** [comment]
> "at some point"? WTF, we already have this! That's the whole point of this.

First if you guys don't explain to us how things work how on earth are we to know?

And secondly, shall I dig up the discussions we had about this where you made me out to be an idiot for even wanting this?

**Comment by tombentley** [comment]
>  we already have this! That's the whole point of this.

Actually, there are improvements we could make in this area. The problem right now is that the SL can say "Oh, I see you're trying to serialize a `X`. `X` is not `serializable`, but I know I can create one by calling `x(String)`", and it can emit something, and then call `x()` during deserialization and it should all work, but it's necessarily a per-SL thing. It could be improved by making it easier for SLs to know about and use this pattern, rather than having to build in special cases. A good example is things like URIs which a better serialized as a single String than their components.

**Comment by gavinking** [comment]
@quintesse the "object that can serialize objects that don't know how the serialize themselves" is called a "serialization library".

**Comment by quintesse** [comment]
> the "object that can serialize objects that don't know how the serialize themselves" is called a "serialization library".

I sincerely hope you're not suggesting that the library should hard-code all those special cases instead of letting you register handlers for types.

**Comment by gavinking** [comment]
> I sincerely hope you're not suggesting that the library should hard-code all those special cases instead of letting you register handlers for types.

Why on earth would I suggest that a library should have an inflexible hard-coded design?

Does Hibernate have a design like that? Does any other serialization library you know have a design like that? Why would _I_ be advocating such a design?

**Comment by quintesse** [comment]
So in fact it's not the "serialization library" that knows how to do it , but the handlers registered with it, okay then why don't you say so.

**Comment by quintesse** [comment]
@tombentley wouldn't it be possible to do this at the level of the intermediary format? So handlers that would know how to turn an instance of type `X` into a map of its properties? Then it could be used by all SLs, right? Then you'd only need SL-specific handlers for those types that really need some format-specific handling.
